File: npm-core/openplugincore/src/openpluginCompletion.ts
import 'isomorphic-fetch';
import { fetchChatCompletion } from './util/fetching';
import { OpenPlugin } from './openPlugin';

type ChatgptFunctionMessage = {
  role: 'function';
  name: string;
  content: string;
};

export async function openpluginCompletion(
  prompt: string,
  pluginName: string | undefined = undefined,
  rootUrl: string | undefined = undefined,
  openaiApiKey: string | undefined = process.env.OPENAI_API_KEY,
  chatgptArgs: object
) {
  if (!openaiApiKey) {
    throw new Error(
      'openaiApiKey not found. You can pass in the parameter openaiApiKey. You can also set the environment variable openaiApiKey=<API-KEY>.'
    );
  }

  process.env.OPENAI_API_KEY = openaiApiKey;
  if (!pluginName && !rootUrl) {
    console.log('NO PLUGIN NAME OR ROOT URL');
    const res = await fetchChatCompletion({
      ...chatgptArgs,
      messages: [{ role: 'user', content: prompt }],
    });
    const json = await res.json();
    return json;
  }

  const plugin = new OpenPlugin(pluginName, rootUrl, openaiApiKey);
  await plugin.init();
  let functionResponse: ChatgptFunctionMessage;
  try {
    functionResponse = await plugin.fetchPlugin({
      prompt: prompt,
      ...chatgptArgs,
    });
  } catch (e: any) {
    if (
      e instanceof Error &&
      (e.message.includes('Not a plugin function') ||
        e.message.includes('No function_call'))
    ) {
      const res = await fetchChatCompletion({
        ...chatgptArgs,
        messages: [{ role: 'user', content: prompt }],
      });
      const json = await res.json();
      return json;
    } else {
      throw e;
    }
  }
  const allChatgptArgs = {
    ...chatgptArgs,
    messages: [{ role: 'user', content: prompt }, functionResponse],
  };
  const summarize = await fetchChatCompletion(allChatgptArgs);
  const summarizeJson = await summarize.json();
  return summarizeJson;
}

File: pypi-core/openplugincore/openplugin_completion.py
from ast import List
import os
import openai
from dotenv import load_dotenv
from .openplugin import OpenPlugin
load_dotenv()

OPENAI_API_KEY = os.getenv('OPENAI_API_KEY')

def openplugin_completion(openai_api_key: str, messages: list[dict], plugin_name: str = None, root_url: str = None, **chatgpt_args):
    # set environment variable to 
    os.environ["OPENAI_API_KEY"] = openai_api_key
    openai.api_key = openai_api_key
    if not plugin_name and not root_url:
        return openai.ChatCompletion.create(
            **chatgpt_args,
            messages=messages
        )
    plugin = OpenPlugin(plugin_name=plugin_name, root_url=root_url, openai_api_key=openai_api_key)
    try:
        function_response = plugin.fetch_plugin(
            messages=messages,
            **chatgpt_args
        )
    except ValueError as e:
        if "Not a plugin function" in str(e):
            return openai.ChatCompletion.create(
                **chatgpt_args,
                messages=messages
            )
        else:
            raise e
    all_chatgpt_args = {
        **chatgpt_args,
        "messages": messages + [function_response]
    }
    summarize = openai.ChatCompletion.create(**all_chatgpt_args)
    return summarize